VocaDbWeb.Core/Controllers/ControllerBase.cs

```diff
@@ -0,0 +1,254 @@
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Net.Http.Headers;
+using Newtonsoft.Json;
+using NLog;
+using VocaDb.Model;
+using VocaDb.Model.DataContracts;
+using VocaDb.Model.DataContracts.UseCases;
+using VocaDb.Model.Domain;
+using VocaDb.Model.Domain.Images;
+using VocaDb.Model.Domain.Security;
+using VocaDb.Model.Domain.Web;
+using VocaDb.Model.Helpers;
+using VocaDb.Model.Utils;
+using VocaDb.Web.Code;
+using VocaDb.Web.Helpers;
+
+namespace VocaDb.Web.Controllers
+{
+	// TODO: implement
+	public class ControllerBase : Controller
+	{
+		private static readonly Logger s_log = LogManager.GetCurrentClassLogger();
+		protected static readonly TimeSpan ImageExpirationTime = TimeSpan.FromMinutes(5);
+		protected const int EntriesPerPage = 30;
+		protected const int InvalidId = 0;
+		protected static readonly TimeSpan PictureCacheDuration = TimeSpan.FromDays(30);
+		protected const int PictureCacheDurationSec = 30 * 24 * 60 * 60;
+		protected const int StatsCacheDurationSec = 24 * 60 * 60;
+
+		protected ControllerBase()
+		{
+			PageProperties.OpenGraph.Image = VocaUriBuilder.StaticResource("/img/vocaDB-title-large.png");
+		}
+
+		protected PagePropertiesData PageProperties => PagePropertiesData.Get(ViewBag);
+
+		protected IUserPermissionContext PermissionContext => HttpContext.RequestServices.GetRequiredService<IUserPermissionContext>();
+
+		private Login Login => HttpContext.RequestServices.GetRequiredService<Login>();
+
+		protected string GetHostnameForValidHit()
+		{
+			return WebHelper.IsValidHit(Request) ? WebHelper.GetRealHost(Request) : string.Empty;
+		}
+
+		protected ActionResult NoId()
+		{
+			return NotFound("No ID specified");
+		}
+
+		protected void AddFormSubmissionError(string details)
+		{
+			s_log.Warn("Form submission error: {0}", details);
+			ModelState.AddModelError(string.Empty, $"Error while sending form contents - please try again. Diagnostic error message: {details}.");
+		}
+
+		protected ActionResult Picture(EntryForPictureDisplayContract contract)
+		{
+			ParamIs.NotNull(() => contract);
+
+			var cacheControl = new CacheControlHeaderValue
+			{
+				// Allow images to be cached by public proxies, images shouldn't contain anything sensitive so this should be ok.
+				Public = true,
+			};
+
+			Response.Headers[HeaderNames.ETag] = $"{contract.EntryType}{contract.EntryId}v{contract.Version}";
+
+			// Cached version indicated by the "v" request parameter.
+			// If no version is specified, assume no caching.
+			if (contract.Version > 0 && !string.IsNullOrEmpty(Request.Query["v"]))
+				cacheControl.MaxAge = PictureCacheDuration;
+
+			Response.GetTypedHeaders().CacheControl = cacheControl;

```

Comment:
I'm not 100% sure if this is the right way to implement this method.

Reply:
Looks okay...